spark/v3.2/spark/src/jmh/java/org/apache/iceberg/spark/action/IcebergSortCompactionBenchmark.java

```diff
@@ -0,0 +1,348 @@
+            .sortBy("timestampCol", SortDirection.DESC, NullOrder.NULLS_FIRST)
+            .sortBy("doubleCol", SortDirection.DESC, NullOrder.NULLS_FIRST)
+            .sortBy("longCol", SortDirection.DESC, NullOrder.NULLS_FIRST)
+            .build())
+        .execute();
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void zSortInt() {
+    SparkActions.get()
+        .rewriteDataFiles(table())
+        .zOrder("intCol")
+        .execute();
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void zSortInt2() {
+    SparkActions.get()
+            .rewriteDataFiles(table())
+            .zOrder("intCol", "intCol2")
+            .execute();
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void zSortInt3() {
+    SparkActions.get()
+            .rewriteDataFiles(table())
+            .zOrder("intCol", "intCol2", "intCol3")
+            .execute();
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void zSortInt4() {
+    SparkActions.get()
+            .rewriteDataFiles(table())
+            .zOrder("intCol", "intCol2", "intCol3", "intCol4")
+            .execute();
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void zSortString() {
+    SparkActions.get()
+        .rewriteDataFiles(table())
+        .zOrder("stringCol")
+        .execute();
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void zSortFourColumns() {
+    SparkActions.get()
+        .rewriteDataFiles(table())
+        .zOrder("stringCol", "intCol", "dateCol", "doubleCol")
+        .execute();
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void zSortSixColumns() {
+    SparkActions.get()
+        .rewriteDataFiles(table())
+        .zOrder("stringCol", "intCol", "dateCol", "timestampCol", "doubleCol", "longCol")
+        .execute();
+  }
+
+  protected Configuration initHadoopConf() {
+    return new Configuration();
+  }
+
+  protected final void initTable() {
+    Schema schema = new Schema(
+        required(1, "longCol", Types.LongType.get()),
+        required(2, "intCol", Types.IntegerType.get()),
+        required(3, "intCol2", Types.IntegerType.get()),
+        required(4, "intCol3", Types.IntegerType.get()),
+        required(5, "intCol4", Types.IntegerType.get()),
+        required(6, "floatCol", Types.FloatType.get()),
+        optional(7, "doubleCol", Types.DoubleType.get()),
+        optional(8, "dateCol", Types.DateType.get()),
+        optional(9, "timestampCol", Types.TimestampType.withZone()),
+        optional(10, "stringCol", Types.StringType.get()));
+
+    SparkSessionCatalog catalog = null;
+    try {
+      catalog = (SparkSessionCatalog)
+          Spark3Util.catalogAndIdentifier(spark(), "spark_catalog").catalog();
+      catalog.dropTable(IDENT);
+      catalog.createTable(IDENT, SparkSchemaUtil.convert(schema), new Transform[0], Collections.emptyMap());
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void appendData() {
+    Dataset<Row> df = spark().range(0, NUM_ROWS * NUM_FILES, 1, NUM_FILES)
```

Comment:
Not sure the mocked data is good to verify the performance. In real world, it always random and exists duplicate data, and it would be more slower during comparing.

FYI, order by binary column is very slow for the duplicate data, https://github.com/apache/spark/pull/34310 fixed it in Spark-3.3.0. It changes comparison to long-by-long rather byte-by-byte.

Reply:
What would you suggest? Here is an example where every single record is the same (probably some good branch prediction helping us out here)

Iteration   1: 237.364 s/op
Iteration   2: 196.935 s/op
Iteration   3: 203.385 s/op

I feel like we have lots of micro benchmarks here, but none that really show that in this context it matters that much.

Reply:
thank you @RussellSpitzer yea, I agree it really depends on the length of ZOrder value. Just post the thought of potential issue.

> What would you suggest

I thought about that do we need a unit benchmark for the method of `ZOrderByteUtils.interleaveBits` ? it is directly and deos not depend on the Spark version or something else.

Reply:
I do think we need lots of benchmarking when we want to tune this up. When Spark 3.0 comes out it will be great to rerun this as well and see if we get better performance. I am 100% in favor of any followups to improve things but I do want to get this out as soon as we can so we have at least a starting point.

Reply:
Ran some benchmarks on ZOrderByteUtils, see the newly attached commit for the test file
Times 10 Million interleaves


ZOrderByteUtilsBenchmark.interleaveValuesFourColumns               ss    5  10.706 ± 1.117   s/op
ZOrderByteUtilsBenchmark.interleaveValuesFourColumns8ByteOutput    ss    5   2.854 ± 0.832   s/op
ZOrderByteUtilsBenchmark.interleaveValuesThreeColumns              ss    5   7.960 ± 1.532   s/op
ZOrderByteUtilsBenchmark.interleaveValuesTwoColumns                ss    5   5.618 ± 1.265   s/op

Reply:
sorry for the late, thank you @RussellSpitzer , looks good to me